src/contracts/collections/positions.rs

```diff
@@ -46,54 +45,200 @@ impl Positions {
+                .take((account_id, positions_length - 1))
+                .unwrap();
+            self.positions.insert((account_id, index), &last_position);
         } else {
-            Err(InvariantError::PositionNotFound)
+            self.positions.remove((account_id, index));
         }
+
+        self.positions_length
+            .insert(account_id, &(positions_length - 1));
+
+        Ok(position.unwrap())
     }
 
     pub fn transfer(
         &mut self,
         account_id: AccountId,
         index: u32,
-        receiver: AccountId,
+        receiver_account_id: AccountId,
     ) -> Result<(), InvariantError> {
         let position = self.remove(account_id, index)?;
-        self.add(receiver, position);
+        self.add(receiver_account_id, position);
 
         Ok(())
     }
 
     pub fn get_all(&self, account_id: AccountId) -> Vec<Position> {
-        self.positions.get(&account_id).unwrap_or_default().1
+        let mut positions = vec![];
+
+        for index in 0..self.get_length(account_id) {
+            let position = self.positions.get((account_id, index)).unwrap();
+            positions.push(position);
+        }
+
+        positions
     }
 
     pub fn get(&mut self, account_id: AccountId, index: u32) -> Option<Position> {
-        let positions_length = self.get_length(account_id);
+        let position = self.positions.get((account_id, index));
+        position
+    }
 
-        if index < positions_length {
-            let positions = self.get_all(account_id);
-            positions.get(index as usize).cloned()
-        } else {
-            None
-        }
+    pub fn get_length(&self, account_id: AccountId) -> u32 {
+        let positions_length = self.positions_length.get(account_id).unwrap_or(0);
+        positions_length
     }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
 
-    fn get_length(&self, account_id: AccountId) -> u32 {
-        let positions = self.get_value(account_id);
-        positions.0
+    #[ink::test]
+    fn test_add() {
+        let positions = &mut Positions::default();
+        let account_id = AccountId::from([0x01; 32]);
+        let position = Position::default();
+
+        positions.add(account_id, position);
+
+        assert_eq!(positions.get(account_id, 0), Some(position));
+        assert_eq!(positions.get(account_id, 1), None);
     }
 
-    fn get_value(&self, account_id: AccountId) -> (u32, Vec<Position>) {
-        let positions = self.positions.get(&account_id).unwrap_or_default();
-        positions
+    #[ink::test]
+    fn test_update() {
+        let positions = &mut Positions::default();
+        let account_id = AccountId::from([0x01; 32]);
+        let position = Position::default();
+        let new_position = Position {
+            lower_tick_index: -1,
+            upper_tick_index: 1,
+            ..Position::default()
+        };
+
+        positions.add(account_id, position);
+
+        let result = positions.update(account_id, 0, &new_position);
```

Comment:
you can use `unwrap` in tests. Asserts in such cases only add irrelevant lines to the code.
`assert_eq!(result, Ok(()));`